New Tag picker implementation

cdharris opened the pull request:
This PR is should undergo final testing before being merged :)

## Done Recently:
- [x] Display 20 most recently used tags in tag picker instances
- [x] Update designs for tag picker in annotations edit mode
- [x] Make sure to close the tag picker when clicking outside of it or pressing ESC
- [x] Add a tag icon to the action items

## Done Previously:
- [x] Replace dashboard results item implementation
- [x] Replace & Fix annotation tag picker (minimally, deferring side bar work).
- [x] Pressing Comma or Tab adds a tag
- [x] use placeholder text "Search & add tags" when the selected list is empty
- [x] Fix failing tests
- [x] Add 'tag all open pages' icon to 'Create new tag' row 
- [x] Add 'tag all open pages' icon to a tag row in the result even when it is already selected (applied to that page)
- [x] Make the 'tag all open pages' button work (just needs hooking up)
- [x] React to backend errors in saving (undo accordingly)
- [x] Add an E2E test with storage layer checking

## Pending:

#### Annotation Tags
_Agreed to: Wait for after Sidebar refactor as not to duplicate work._
- [x] Annotation tag picker (UX) - replace component to be more consistent, following the Figma mockup
- [x] Annotation tag picker (bug) - does not load all existing tags. reproduce by editing the tags of an existing annotation. just see the ones there that have already been selected.

#### Filter Picker
_Recommendation: Include as part of Collection Picker work due to difference in logic._
_See more in comment below_

#### UX improvements
- [ ] Delete to remove a tag
    (What about when holding down delete to clear search input, should it immediately start deleting selected tags?)

## Next
My preference would be for this PR to tested and pending any last fixes that are not already accounted for in the above list, to be merged into develop.

This will allow us to scope and prioritise these other tangential issues independently (tag filtering & collections, tag suggestions), factoring them into the overall product release roadmap.

If agreed I will create new issues for these other tasks and scope them out there.

blackforestboi commented:
## Review 30.03.2020

- [ ] don't show selected tags in both the list and the top bar. Just show them in the top bar. 
- [ ] when clicking on a tag to select it, keep the cursor in the text field so that a user can continue typing
- [ ] when using the keyboard to select items (so not yet selected), dont have them have the same style than a fully selected. Very confusing. 
- [ ] Don't show the checkmark when hovering over an item (keyboard and mouse), simply show the "tag all open tabs" button.

kellective commented:
There are two different looking styles between the top accessed tag picker and the ribbon accessed tag picker (slightly) which is weird because these should be using the same components.

### Ribbon
<img width="327" alt="Screen Shot 2020-03-31 at 7 58 11 PM" src="https://user-images.githubusercontent.com/121013/77997146-6461c980-738b-11ea-8fc7-b488859edd24.png">

### Top
<img width="318" alt="Screen Shot 2020-03-31 at 7 58 24 PM" src="https://user-images.githubusercontent.com/121013/77997157-675cba00-738b-11ea-9129-19114b8d9532.png">


Styling things I'll look into:
[ ] Make consistent font sizes and font weights
[ ] Lighten up the icon colour
[ ] Style the back button

Chris, when I add a tag sometimes it makes a new row even if there is space to the right of an existing tag for it to sit next to. I'm unsure how to fix this. I think it might be fiddly but i can keep trying.
<img width="303" alt="Screen Shot 2020-03-31 at 7 39 46 PM" src="https://user-images.githubusercontent.com/121013/77997210-7fccd480-738b-11ea-8e27-90cb64df227e.png">
<img width="355" alt="Screen Shot 2020-03-31 at 8 06 05 PM" src="https://user-images.githubusercontent.com/121013/77997227-86f3e280-738b-11ea-8830-9b9c804a56ba.png">

cdharris commented:
> when I add a tag sometimes it makes a new row even if there is space to the right of an existing tag for it to sit next to. 
I think this is due to the default `size` attribute of the `<input />` element, it seems to be 20. and will default to that size. Not sure how best to deal with this for now, other than making sure when you click the search box in general it focuses the input.

blackforestboi commented:
@cdharris in the call we talked about that the selected tags on top of the bottom list are only on top because they have been recently selected, and would be moved down if other tags are selected on other pages. I can confirm that this is NOT the case. 
They are the tags selected on that page and stay on top no matter if I tag other tags on other pages. They should be the last added/used tags.

cdharris commented:
> @cdharris in the call we talked about that the selected tags on top of the bottom list are only on top because they have been recently selected, and would be moved down if other tags are selected on other pages. I can confirm that this is NOT the case.
> They are the tags selected on that page and stay on top no matter if I tag other tags on other pages. They should be the last added/used tags.
> 
> EDIT: Here are also the updated figma files: https://www.figma.com/file/Ec1NJKz82yAJ2MEXipap0K/Tags?node-id=28%3A549



> @cdharris in the call we talked about that the selected tags on top of the bottom list are only on top because they have been recently selected, and would be moved down if other tags are selected on other pages. I can confirm that this is NOT the case.
> They are the tags selected on that page and stay on top no matter if I tag other tags on other pages. They should be the last added/used tags.
> 
> EDIT: Here are also the updated figma files: https://www.figma.com/file/Ec1NJKz82yAJ2MEXipap0K/Tags?node-id=28%3A549



> They are the tags selected on that page and stay on top no matter if I tag other tags on other pages. They should be the last added/used tags.

Can I confirm, you're talking about the order of the tags within the results list here?
You're saying the order of default tags shown (tag suggestions) should be descending in order of most recently added/tagged ? and that search results should remain alphabetical?

blackforestboi commented:
> Can I confirm, you're talking about the order of the tags within the results list here?
You're saying the order of default tags shown (tag suggestions) should be descending in order of most recently added/tagged ? and that search results should remain alphabetical?

yes I am talking about the order of the result list (default or searched)
Default tags should be the recently used/added tags with the most recent ones on top. 
Search results in alphabetical order make sense yeah.

kellective commented:
Hi guys,

I did some QA testing today and have found a few errors.

**1. Hover tooltips in sidebar**
Possibly not related but I've noticed it here and not in 2.0.13
_What is happening:_ 
_Expected:_ Current release behaviour (they close when you are no longer hovering on the icon)
The hover tooltips on the sidebar now don't close when the popups are open and focused on. Two examples:
<img width="332" alt="Screen Shot 2020-04-06 at 3 16 07 PM" src="https://user-images.githubusercontent.com/121013/78520734-aa031400-781b-11ea-9b37-c3958cbdb048.png">
<img width="317" alt="Screen Shot 2020-04-06 at 3 16 11 PM" src="https://user-images.githubusercontent.com/121013/78520730-a8395080-781b-11ea-95d6-f1a5a8949a31.png">

**2. Clicking anywhere in the search area should focus the cursor there**. 
_What is happening:_ 
Currently if i click in this empty space i expect to focus the search text and be able to start typing  (ignore the green bar i just highlighted it during testing so i could see where the input field was)
![Frame 3](https://user-images.githubusercontent.com/121013/78520806-edf61900-781b-11ea-9e98-a754302d0aa0.png)

3. Tags don't seem to be saving
_What is happening:_
Clicking on the tag row doesn't seem to se 'saving' the tag. It adds it to the search area but if i close the popup and reopen or go look at my dashboard it i expect the tags i've applied to be there and they aren't. _However_ if i click the 'tag all tabs' button instead of the actual row then they correctly save (to all tabs). So the 'add to this page only' doesn't seem to be working.

_To replicate:_ 
Access the popup (top bar)
Click 'Add tags'
Click a tag row, see the tag get applied. Hit 'Back' 
Click 'Add tags' again, are you tags still applied?

Try this again but instead of click the row click the 'all tabs' button. This should work.

Similar situation with using the enter to add a tag from the search field.

The following errors are produced:
Background_page
<img width="736" alt="Screen Shot 2020-04-06 at 3 42 16 PM" src="https://user-images.githubusercontent.com/121013/78521203-41b53200-781d-11ea-9dcb-a446fc360c68.png">
Console
<img width="547" alt="Screen Shot 2020-04-06 at 3 43 51 PM" src="https://user-images.githubusercontent.com/121013/78521296-7c1ecf00-781d-11ea-829a-7877948f0dee.png">

**4. Icons flicker on hover.** 
I can't seem to work this out, either they not there at all and show only when selected or they are there on row hover but flicker rapidly when the mouse goes over the icon. @cdharris we spoke about this over the phone but i'm still not able get all the conditions working at the same time. 

I'm happy to change the way these are supposed to be but I'm going off the figma file which is:
X icon only visible when tag is selected and row is hovered. 
Tabs icon visible when row is hovered and stays visible when selected.

cdharris commented:
Hi Kelly, thanks for giving this a round of QA testing! Appreciated!

2. Fixed in ee33d82ede0429acb1964cec1cfd8b9bfdd36cd9

3. I can't reproduce this. 

 Adding/Removing/Creating tags from the popup or the ribbon are saving for me. Your error suggested something to do with the page indexing. So I turned off those settings and tried to visit new pages - everything still worked as expected (tag changes reflected when opening the popup or sidebar again and showing up respectively in the dashboard).
I'm wondering if there was something failing with indexing the specific page you were trying to tag? (what was the url? - trying out random wikipedia pages works for me). Or otherwise, maybe some development extension state issues (removing the extension and reinstalling might work.)

 Either way what this brings up is the related issue - to make sure the oppertunistic updates we perform get 'undone' if there are backend failures like in your case. So it's not the situation that you click add, it fails, but it looks like it worked. Fixed: c98a49553f5048c87ca9f65cb1f8562539e24fa2

4. I've reverted the hover implmentation to my last commit in order to fix the flickering issue. 810d0ea12eefffca82373823147099420f5c110f

 The reason I used a property of the element to conditionally switch the CSS opacity property in this case rather than conditionally render the element is that it's generally faster to toggle CSS properties than it is to create new elements and cause greater dom layout updates.

 In this case there seemed to be some conflicting events due to the use of MouseOver/Out also applying to child elements, which works a bit differently from the more common MouseEnter/Leave, but the latter wasn't working with the ribbon, likely a react/shadowdom issue, so I used MouseOver/Out here. But don't worry about those intricate details, just FYI. Let me know if the styling isn't right on the hover and we can tweak it.

cdharris commented:
Lots of final polish on the tag filter implementation.

(I've updated the respective checkboxes in the parent comments.)
**Now done:**
 - 'Add tag to all tabs' button is now included in the create new tag row also.
- Clicking anywhere on the search area focuses the input, and focus should always be on the input after a UI interaction.
- Fixed some small style issues with the flickering icons
- Cleaned up the implementations of this new tag picker, removed the commented out old one.
- If the upstream storage operation fails, the UI change is undone (e.g. adding,creating or removing a tag from a page)
- Added end to end tests that ensure the underlying storage operation works
- Added tests to ensure UI is in the expected state after an error.

Just the issue with the **suggested tags** that I feel is the final point left to decide whether to resolve here or in another issue.

The implementation hasn't been changed, so it's doing what it was before which is:

Loading the redux state with tag suggestions on page loads via `SuggestPlugin.suggestExtended` on `src/search/plugins/suggest.ts:120` The query seems to bring the tags back without any consistent ordering.
Then the redux state does some ordering based on tag adding or deleting, but this appears to be local to the tab you're in.

I think it might be best to revist this with the refactor to the sidebar and included efforts on change watching / state sharing). Alternatively, I can spec out a proposal and implementation within this effort, but it will delay merging this feature in a little bit.

blackforestboi commented:
Agreed, lets fix the suggested tags in a separate issue but asap. 

there are tests failing though, would be great to have them all work so I can merge into develop if everything works and get a release candidate on the road that we can do the last fixes on. 

So happy about this 😍

cdharris commented:
Actually Kelly to your point of  3. Tags don't seem to be saving, in my final run through today and double checking with reloading the extension after turning off all the page indexing settings, I do get a (not the same but) similar error. Did you have page indexing turned offf by any chance? I've pushed a fix for the issue I found.

blackforestboi commented:
Ok so I merged `develop` now into this branch to get it up to speed, hope I did everything all right. 

There are a couple of issues to resolve in priority
- When building this branch ig et the following error: <img width="654" alt="Screenshot 2020-04-06 at 22 49 18" src="https://user-images.githubusercontent.com/7870039/78606807-32a2a400-785e-11ea-820c-120f8711e45a.png">
- replace tag picker in dashboard page results
- Replace tag picker in new annotation/note card <img width="467" alt="Screenshot 2020-04-06 at 22 59 14" src="https://user-images.githubusercontent.com/7870039/78606918-67aef680-785e-11ea-9a4b-c09d797ba518.png">
- replace tag picker in annotation edit mode. Make it so that when clicking on the tag the tag picker pops up as an external box, just like everywhere else too, not inside the annotation card, update the tags in the card when the box closes <img width="455" alt="Screenshot 2020-04-06 at 22 59 44" src="https://user-images.githubusercontent.com/7870039/78606991-88774c00-785e-11ea-9feb-346997974441.png">
- the bottom bar in the popup is entirely missing, no settings reachable anymore <img width="319" alt="Screenshot 2020-04-06 at 23 31 06" src="https://user-images.githubusercontent.com/7870039/78607099-b52b6380-785e-11ea-8db2-2641a6194d27.png">
- when i click on "tag all open tabs" in the tag picker in the ribbon in a tag picker that is not assigned yet to the page, it is not added to the selected items in the UI (it is though in the background) https://recordit.co/op60St5Xuh
- make the tags that are on the top being the last ones used systemwide
- remove the "tag all open tabs/collections" from the popup main menu, confusing why its there twice.

kellective commented:
- [x ] the bottom bar in the popup is entirely missing, no settings reachable anymore
^^ whoops my bad. Fixed!

kellective commented:
@cdharris i think i had all setting on, it was a brand new test env i had set up and i can't imagine i'd have played with any settings. Sorry i can't confirm that for you as I reloaded my extension today to make sure i was testing the latest work.

I can say though that i am no longer getting this error!

kellective commented:
There was 1 todo i had in there (i just removed the TODO note from the code) but this is to 

- [ ] use placeholder text "Search & add tags" when the selected list is empty

kellective commented:
Today i tested in FF and noticed the search box sits below the icon, i'll look into this (green area)
<img width="306" alt="Screen Shot 2020-04-10 at 7 12 00 PM" src="https://user-images.githubusercontent.com/121013/78971318-6718a780-7b5f-11ea-902d-0e639a67dd81.png">

blackforestboi commented:
### Review 13.04.2020
- [ ] Adding to the known bug I found that the annotation tag picker does not load all existing tags. reproduce by editing the tags of an existing annotation. just see the ones there that have already been selected. 

**From the last review these are still outstanding:** 
- [ ] replace tag picker in dashboard page results (show also there the "tag all open tabs" icon)<img width="681" alt="Screenshot 2020-04-13 at 18 19 14" src="https://user-images.githubusercontent.com/7870039/79137673-69733f80-7db3-11ea-8195-8daa4b59c77c.png">
- [ ] replace tag picker in annotation edit mode. Make it so that when clicking on the tag the tag picker **pops up as an external box**, just like everywhere else too, not inside the annotation card, update the tags in the card when the box closes. Is that unclear?

We're almost there 😍

blackforestboi commented:
@cdharris 

Here is the mockups to make the collections picker also in the style of the new tag picker. 
Essentially nothing really changes except the styles for the list of collections. 
https://www.figma.com/file/Ec1NJKz82yAJ2MEXipap0K/Tags?node-id=380%3A0
I feel we should try to fully finish this work before the product hunt release so we don't have to touch it again. 

Meaning 4 things are still necessary: 
- [ ] Tag/collections suggestions to display the 20 last used 
- [ ] Use 'comma' or 'tab' to add new items
- [ ] replace collections picker in ribbon and popup, and add to dashboard
- [ ] fix annotation box styles
- [ ] clicking outside of the annotation-tag-edit box should close the tag picker.
- [ ] clicking on the "note" icon in a dashboard result item causes a "Cannot read property 'initTagSuggestions' of undefined" (got one annotation with a tag on that one)

Prio 2: (not necessary for launch)
- [ ] and 'backspace' to delete items (i reckon that may create some more engineering efforts)

### Styles to fix
These are more things that we still need to take care of to finalise everything, not something @cdharris needs to worry about: 
- [ ] tags that are too long need to cut off: <img width="334" alt="Screenshot 2020-04-17 at 13 12 34" src="https://user-images.githubusercontent.com/7870039/79563450-1fd56e00-80ad-11ea-9fbf-eaaae6cd0f76.png"> <img width="347" alt="Screenshot 2020-04-17 at 13 12 59" src="https://user-images.githubusercontent.com/7870039/79563475-2ebc2080-80ad-11ea-9241-7fbca546e6d3.png">
- [ ] search field to small: <img width="331" alt="Screenshot 2020-04-17 at 13 14 22" src="https://user-images.githubusercontent.com/7870039/79563605-63c87300-80ad-11ea-8a7a-324462d2fdb5.png">

kellective commented:
@blackforestboi i fixed the tag word wrap issue. What do you mean by search field too small? 

Also i updated the old styles of the tags (how they appear on the dashboard content for example) so they all have the new font, padding size etc. They aren't upgraded to styled components but I've just updated the TagPill css for them.

cdharris commented:
# Tag Picker

## Done Recently:

- [x] Replace dashboard results item implementation
- [x] Replace & Fix annotation tag picker (minimally, deferring side bar work).
- [x] Pressing Comma or Tab adds a tag

## Pending:

#### Annotation Tags
_Agreed to: Wait for after Sidebar refactor as not to duplicate work._
- [ ] Annotation tag picker (UX) - replace component to be more consistent, following the Figma mockup
- [ ] Annotation tag picker (bug) - does not load all existing tags. reproduce by editing the tags of an existing annotation. just see the ones there that have already been selected.

#### Filter Picker
_Recommendation: Include as part of Collection Picker work due to difference in logic._
- Filter picker - Replace instances of adding/removing filters with new tag picker, extend tag picker logic as needed

    This involves:
    - [ ] Allow different buttons to be used on the tag picker row (plus, minus)
    - [ ] Use different logic for this since they both 'add' a tag to the search input, in either included or excluded state.
    - [ ] There's no ability to add a new tag

    Options I see for this are between:

    1) Adding this functionality to the existing tag picker, behind conditional properties. Starts to mix up the logic - not preferred, it's one of the things we were seeking to get rid of.

    2) Copying the TagPicker and modifying it to be a TagFilter - the fastest and cleaner than the 1st option but would have a fair amount of duplication of code.

    3) Extracting out the generic components that can be composed to reduce duplication between TagPicker and TagFilter, this will end up with a component that is cleaner and contains less duplications, but because of the additional moving parts, may be a bit more involved to read.

    Verdict: To aim for 3., falling back to 2.  with elements of 3. And doing this work as part of the Collection Picker work, as it will be a nice segway into extracting the TagPicker logic to be more general and composeable.


### UX improvements

- [ ] Delete to remove a tag

    (What about when holding down delete to clear search input, should it immediately start deleting selected tags?)


## Next

My preference would be for this PR to tested and pending any last fixes that are not already accounted for in the above list, to be merged into develop.

This will allow us to scope and prioritise these other tangential issues independently (tag filtering & collections, tag suggestions), factoring them into the overall product release roadmap.

If agreed I will create new issues for these other tasks and scope them out there.

blackforestboi commented:
We are almost ready to merge this in! 

There is only one thing which is odd really: The way the tag picker works in the "annotation-edit" dialogue. 
Right now it looks like this. 
<img width="439" alt="Screenshot 2020-04-20 at 19 25 35" src="https://user-images.githubusercontent.com/7870039/79781334-9ee0d580-833d-11ea-80a3-87f41512ceb7.png">

But in the end it should look more like this: https://www.figma.com/file/Ec1NJKz82yAJ2MEXipap0K/Tags?node-id=392%3A0 

Also its important to close the tag picker when a user clicks outside or clicks ESC (but dont close the entire sidebar with ESC, when tag picker is open) or it otherwise the UX is hard to understand (clicking again on the tag field) to close it. 

So todos: 
- [ ] update designs for tag picker in annotations edit mode
- [ ] make sure to close the tag picker when clicking outside of it or pressing ESC
- [ ] add a tag icon to the action items

cdharris commented:
> update designs for tag picker in annotations edit mode

Done - with the slight difference that the picker is loaded below the row of selected tags. There was some fiddly css to get it right otherwise, perhaps Oli or Kelly can have a look here.

> make sure to close the tag picker when clicking outside of it or pressing ESC

Done

> add a tag icon to the action items

Done. (I think I understood this right - replacing the plus symbol for the tag icon)

kellective commented:
Hey chris, just on your last comment about the icon, i havent checked your change but i believe the request was about the idea we'd add 'tags' as a clickable option from within the annotation action buttons. We discussed briefly that we like it but i don't know if we decided whether it was in scope or what needs to change to make it work from there (adding tags while not being in edit mode)

https://www.figma.com/file/Ec1NJKz82yAJ2MEXipap0K/Tags?node-id=366%3A1

cdharris commented:
> add 'tags' as a clickable option from within the annotation action buttons.

Ah thanks for the added clarification. I agree I like this better, but it invoves a bit of sidebar work and I thought we had decided to postpone this.

blackforestboi commented:
> Done. (I think I understood this right - replacing the plus symbol for the tag icon)

Yeah the idea was that a user doesn't have to click edit and then click on the tags to edit tags but just like with the page result item can do that right from the annotation result element. 
But we can also just leave the button out completely. Should be clear that the grey bar indicates adding tags. 

> make sure to close the tag picker when clicking outside of it or pressing ESC

This does not seem to work. https://recordit.co/mW6s0b9hWw 

Expected is that the tag picker closes, and when using ESC it only closes the tag picker overlay, not the sidebar completely. (the latter may involve some more sidebar work though - so another thing to postpone.

kellective commented:
Oh actually...its reference to this. Which will be slightly different because it looks like we dont have the other actions available to us from that mode right now.

<img width="434" alt="Screen Shot 2020-04-21 at 10 40 36 PM" src="https://user-images.githubusercontent.com/121013/79857500-fb1d2580-8421-11ea-8b83-8684ed2b447a.png">

cdharris commented:
Ah, yes thanks for the clarification.
See: https://github.com/WorldBrain/Memex/pull/985#issuecomment-616673938
> ## Pending:
> #### Annotation Tags
> _Agreed to: Wait for after Sidebar refactor as not to duplicate work._
> 
> * [ ]  Annotation tag picker (UX) - replace component to be more consistent, following the Figma mockup

This looks like a simple change but there's a reasonable amount of the old implemenation that needs to be ripped out and replaced to get it to work. We decided it's best to wait till after Vincent's work on the sidebar refactor as not to duplicate this.

cdharris commented:
> we can also just leave the button out completely.
Ok Done

> This does not seem to work. https://recordit.co/mW6s0b9hWw

That's strange I'm getting different results. It looked like somehow your focus was stuck on the Cancel button, but I couldn't replicate that here, tried by tabbing on different elements first.
![tag5](https://user-images.githubusercontent.com/3830518/79858870-5e7e6780-83d0-11ea-9c9d-1a78352a8761.gif)

blackforestboi commented:
ok, lets fix that later then, seems like a rabbit hole. Let's focus on the suggestions first.

blackforestboi commented:
Ok did another testing round. 
The tag suggestions work flawlessly. 

From my side it still needs a few style things that need to be fixed. 
But there is one rather annoying bug that seems to be easy to fix: 
When I remove a tag via the tag picker in the dashboard, the overlay closes. Really annoying if I want to remove many tags. Also it would be good if the overlay closes with the ESC button. 

Also the problem with the tag picker not closing on click or esc in the sidebar annotations is indeed rather annoying and we should fix it before release. I'll put it on the tasks for next week. 

Other than that fantastic update, can't wait to use it more in action.

blackforestboi commented:
Another idea on the annotation tag picker not closing on ESC or click outside. Can you just manually make sure that the element in focus is indeed the tag picker and not anything else?

kellective commented:
- [ ] use placeholder text "Search & add tags" when the selected list is empty
^^ this is marked as done at the top of the PR but itsn't done.  Can we squeeze in the change request for the tag icon and 'Add tags to page' 'Add tags to note' ? :) 
![tags](https://user-images.githubusercontent.com/121013/80347953-2b047700-88c1-11ea-96bd-c910ef019351.png)

blackforestboi commented:
ok @poltak @cdharris. Awesome work to get this done :) 

There are still some tests failing on this branch that we need to take care of - and once the sidebar UI refactor is done the issues with the keyboard and click actions in the annotations tag picker.
